common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,147 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.crypto;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.crypto.tink.subtle.AesGcmJce;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.*;
+import org.apache.spark.network.util.AbstractFileRegion;
+import io.netty.buffer.ByteBuf;
+
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.WritableByteChannel;
+import java.security.GeneralSecurityException;
+
+public class GcmTransportCipher implements TransportCipher {
+    private static final byte[] DEFAULT_AAD = new byte[0];
+
+    private final SecretKeySpec aesKey;
+    public GcmTransportCipher(SecretKeySpec aesKey)  {
+        this.aesKey = aesKey;
+    }
+
+    @VisibleForTesting
+    EncryptionHandler getEncryptionHandler() {
+        return new EncryptionHandler();
+    }
+
+    @VisibleForTesting
+    DecryptionHandler getDecryptionHandler() {
+        return new DecryptionHandler();
+    }
+
+    public void addToChannel(Channel ch) {
+        ch.pipeline()
+            .addFirst("GcmTransportEncryption", getEncryptionHandler())
+            .addFirst("GcmTransportDecryption", getDecryptionHandler());
+    }
+
+    @VisibleForTesting
+    class EncryptionHandler extends ChannelOutboundHandlerAdapter {
+        @Override
+        public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+                throws Exception {
+            ByteBuffer inputBuffer;
+            int bytesToRead;
+            if (msg instanceof ByteBuf byteBuf) {
+                bytesToRead = byteBuf.readableBytes();
+                // This is allocating a buffer that is the size of the input
+                inputBuffer = ByteBuffer.allocate(bytesToRead);
+                // This will block while copying
+                while (inputBuffer.position() < bytesToRead) {
+                    byteBuf.readBytes(inputBuffer);
+                }
+            } else if (msg instanceof AbstractFileRegion fileRegion) {
+                bytesToRead = (int) fileRegion.count();
+                // This is allocating a buffer that is the size of the input
+                inputBuffer = ByteBuffer.allocate(bytesToRead);
+                ByteBufferWriteableChannel writeableChannel =
+                        new ByteBufferWriteableChannel(inputBuffer);
+                long transferred = 0;
+                // This will block while copying
+                while (transferred < bytesToRead) {
+                    transferred +=
+                            fileRegion.transferTo(writeableChannel, fileRegion.transferred());
+                }
```

Comment:
This can be quite large - and result in a nontrivial memory consumption.
We will need to do something similar to `CtrTransportCipher.EncryptedMessage` and transfer in chunks.